runtime/ops/tls.rs

```diff
@@ -70,6 +88,457 @@ impl StoresClientSessions for ClientSessionMemoryCache {
+            // Do a zero-length plaintext read so we can detect the arrival of
+            // 'CloseNotify' messages, even if only the write half is active.
+            // Actually reading data from the socket is done in `poll_read()`.
+            Ok(()) => match self.tls.read(&mut []) {
+              Ok(0) => {}
+              Ok(_) => unreachable!(),
+              // Rustls `read()` returns `ConnectionAborted` when it receives a
+              // 'CloseNotify' alert; this indicates that the remote end of the
+              // connection has initiated a graceful shutdown at the TLS level.
+              Err(err) if err.kind() == ErrorKind::ConnectionAborted => {
+                self.rd_shut = Shut::TlsShut;
+              }
+              Err(err) => return Poll::Ready(Err(err)),
+            },
+            Err(err) => {
+              let err = io::Error::new(ErrorKind::InvalidData, err);
+              return Poll::Ready(Err(err));
+            }
+          },
+          Err(err) if err.kind() == ErrorKind::WouldBlock => {
+            match self.tcp.poll_read_ready(cx) {
+              Poll::Ready(result) => result?,
+              Poll::Pending => tcp_read_ready = false,
+            }
+          }
+          Err(err) => return Poll::Ready(Err(err)),
+        }
+      }
+
+      if tcp_write_ready && self.tls.wants_write() {
+        continue;
+      }
+
+      let poll_again = match flow {
+        _ if self.tls.is_handshaking() => true,
+        Flow::Read => self.tls.wants_read() && self.rd_shut != Shut::TcpShut,
+        Flow::Write => self.tls.wants_write(),
+      };
+      return match poll_again {
+        false => Poll::Ready(Ok(())),
+        true => Poll::Pending,
+      };
+    }
+  }
+
+  fn poll_graceful_close(&mut self, cx: &mut Context<'_>) -> Poll<()> {
+    let p1 = self.poll_io(cx, Flow::Read);
+    let p2 = Pin::new(self).as_mut().poll_shutdown(cx);
+    let _ = ready!(p1);
+    let _ = ready!(p2);
+    Poll::Ready(())
+  }
+}
+
+impl AsyncRead for TlsStream {
+  fn poll_read(
+    mut self: Pin<&mut Self>,
+    cx: &mut Context<'_>,
+    buf: &mut ReadBuf<'_>,
+  ) -> Poll<io::Result<()>> {
+    ready!(self.poll_io(cx, Flow::Read))?;
+
+    if matches!(self.rd_shut, Shut::TlsShut | Shut::TcpShut) {
+      return Poll::Ready(Ok(()));
+    }
+
+    let buf_slice =
+      unsafe { &mut *(buf.unfilled_mut() as *mut [MaybeUninit<u8>] as *mut _) };
+
+    match self.tls.read(buf_slice) {
+      Ok(0) => unreachable!(),
+      Ok(bytes_read) => {
+        unsafe { buf.assume_init(bytes_read) };
+        buf.advance(bytes_read);
+        Poll::Ready(Ok(()))
+      }
+      Err(err) => Poll::Ready(Err(err)),
+    }
+  }
+}
+
+impl AsyncWrite for TlsStream {
+  fn poll_write(
+    mut self: Pin<&mut Self>,
+    cx: &mut Context<'_>,
+    buf: &[u8],
+  ) -> Poll<io::Result<usize>> {
+    ready!(self.poll_io(cx, Flow::Write))?;
+
+    match self.tls.write(buf) {
+      Ok(0) => unreachable!(),
```

Comment:
It may return 0 when the internal buffer of rustls is full.

Reply:
That shouldn't happen, because `poll_io()` return `Poll::Pending` when the buffer is full.
However it will return 0 when `buf` is empty, so I removed the assertion anyway.

Reply:
No, it will not return to pending. see https://github.com/ctz/rustls/blob/47d309233f2d21ccf058d35c58d684e5e4c9c643/rustls/src/server/mod.rs#L738

Reply:
> 
No, it will not return to pending. see https://github.com/ctz/rustls/blob/47d309233f2d21ccf058d35c58d684e5e4c9c643/rustls/src/server/mod.rs#L738

`poll_io()` returns `Poll::Pending` when the buffer is full: https://github.com/denoland/deno/blob/800ae5ee1f60f30649f09e9a2b53599afc663760/runtime/ops/tls.rs#L325

Reply:
Sorry, I didn't see to `poll_io`, you are right.